src/dotty/tools/dotc/core/tasty/TreePickler.scala

```diff
@@ -346,181 +342,226 @@ class TreePickler(pickler: TastyPickler) {
     stats.foreach(stat => if (!stat.isEmpty) pickleTree(stat))
   }
 
-  def pickleTree(tree: Tree)(implicit ctx: Context): Unit = try {
-    registerTreeAddr(tree)
-    tree match {
-      case tree if tree.isType =>
-        pickleTpt(tree)
-      case Ident(name) =>
-        tree.tpe match {
-          case tp: TermRef => pickleType(tp)
-          case _ =>
-            writeByte(IDENT)
-            pickleName(name)
-            pickleType(tree.tpe)
-        }
-      case This(_) =>
-        pickleType(tree.tpe)
-      case Select(qual, name) =>
-        writeByte(SELECT)
-        val realName = tree.tpe match {
-          case tp: NamedType if tp.name.isShadowedName => tp.name
-          case _ => name
-        }
-        val sig = tree.tpe.signature
-        if (sig == Signature.NotAMethod) pickleName(realName)
-        else pickleNameAndSig(realName, sig)
-        pickleTree(qual)
-      case Apply(fun, args) =>
-        writeByte(APPLY)
-        withLength {
-          pickleTree(fun)
-          args.foreach(pickleTree)
-        }
-      case TypeApply(fun, args) =>
-        writeByte(TYPEAPPLY)
-        withLength {
-          pickleTree(fun)
-          args.foreach(pickleTpt)
-        }
-      case Literal(const1) =>
-        pickleConstant {
+  def pickleTree(tree: Tree)(implicit ctx: Context): Unit = {
+    val addr = registerTreeAddr(tree)
+    if (addr != currentAddr) {
+      writeByte(SHARED)
+      writeRef(addr)
+    }
+    else
+      try tree match {
+        case Ident(name) =>
           tree.tpe match {
-            case ConstantType(const2) => const2
-            case _ => const1
+            case tp: TermRef if name != nme.WILDCARD =>
+              // wildcards are pattern bound, need to be preserved as ids.
+              pickleType(tp)
+            case _ =>
+              writeByte(if (tree.isType) IDENTtpt else IDENT)
+              pickleName(name)
+              pickleType(tree.tpe)
           }
-        }
-      case Super(qual, mix) =>
-        writeByte(SUPER)
-        withLength {
-          pickleTree(qual);
-          if (!mix.isEmpty) {
-            val SuperType(_, mixinType) = tree.tpe
-            pickleType(mixinType)
+        case This(_) =>
+          pickleType(tree.tpe)
```

Comment:
The qualifier of a `This` tree might be something we would need (for example when renaming a class), should we pickle the complete tree instead of just pickling its type?